src/OpenTelemetry/SuppressInstrumentationScope.cs

```diff
@@ -86,18 +86,8 @@ public static int Enter()
             return value;
         }
 
-        /// <inheritdoc/>
-        public void Dispose()
-        {
-            if (!this.disposed)
-            {
-                Slot.Set(this.previousValue);
-                this.disposed = true;
-            }
-        }
-
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal static int IncrementIfTriggered()
+        public static int IncrementIfTriggered()
```

Comment:
@ejsmith What if instead of making these two methods public we made it a partial class and moved them to a separate file so that file could be part of the include list? I bring it up just because I don't think it is a very nice API that was ever intended to be public? I think they are only called by a couple of other very internal-y spots. /cc @reyang

Reply:
Maybe won't work because it will think it's a different class? Doh

Reply:
Yeah, pretty sure that wouldn't work.

Reply:
@alanwest @cijothomas @reyang 

It looks like we need to promote `IncrementIfTriggered` & `DecrementIfTriggered` to the public API because they are used by `DiagnosticSourceListener`. Does anyone remember why? I just stepped through the code and it looks like `TracerProviderSdk` also does it, so we're duplicating effort.

Would this work and allow us to keep them internal to SDK?

```csharp
        public void OnNext(KeyValuePair<string, object> value)
        {
            if (!Sdk.SuppressInstrumentation) // Do nothing if suppressing.
            {
                return;
            }

            if (!this.handler.SupportsNullActivity && Activity.Current == null)
            {
                InstrumentationEventSource.Log.NullActivity(value.Key);
                return;
            }

            try
            {
                if (value.Key.EndsWith("Start", StringComparison.Ordinal))
                {
                    this.handler.OnStartActivity(Activity.Current, value.Value);
                }
                else if (value.Key.EndsWith("Stop", StringComparison.Ordinal))
                {
                    this.handler.OnStopActivity(Activity.Current, value.Value);
                }
                else if (value.Key.EndsWith("Exception", StringComparison.Ordinal))
                {
                    this.handler.OnException(Activity.Current, value.Value);
                }
                else
                {
                    this.handler.OnCustom(value.Key, Activity.Current, value.Value);
                }
            }
            catch (Exception ex)
            {
                InstrumentationEventSource.Log.UnknownErrorProcessingEvent(this.handler?.SourceName, value.Key, ex);
            }
        }
```

Reply:
That doesn't work because it causes the `OnStopActivity` to get skipped.

Reply:
In `TracerProviderSdk` it is more like a _nice-to-have_ thing to use `IncrementIfTriggered` & `DecrementIfTriggered` - to prevent instrumentation mistakes (failed to maintain the ref-count).
In `DiagnosticSourceListener` it is needed since this is where the _reentrant-mutex_ like feature was intended for.

We might need to find better names if we want them to be public.

Reply:
This may be able to just be a local flag in the DiagnosticSourceListener. I think we just need to have a flag that says skip the next stop call if start was skipped.

Reply:
Maybe something like this:

```csharp
                if (value.Key.EndsWith("Start", StringComparison.Ordinal))
                {
                    if (!Sdk.SuppressInstrumentation)
                    {
                        this.handler.OnStartActivity(Activity.Current, value.Value);
                    }
                    else
                    {
                        this.skipNextStopCall = true;
                    }
                }
                else if (value.Key.EndsWith("Stop", StringComparison.Ordinal))
                {
                    if (!this.skipNextStopCall)
                    {
                        this.handler.OnStopActivity(Activity.Current, value.Value);
                    }
                    else
                    {
                        this.skipNextStopCall = false;
                    }
                }
```

Reply:
Is `this.skipNextStopCall` thread safe (e.g. if multiple threads are trying to modify/access the same thing)?

Reply:
@reyang I was thinking that it didn't have to be because it's an instance variable for that instance of the diagnostic source listener and as far as I know a single listener isn't multi-threaded.

Reply:
I'm very confused by "a single listener isn't multi-threaded", would you elaborate?

Reply:
I don't know that for sure. You guys would know it better. But I'd say that if they were multithreaded then I'm not sure how the existing counter approach would work. What happens if I get one start call that is suppressed and then one that isn't and then the stop is called for the unsuppressed one before the original one that is suppressed?

Reply:
It seems like it might be better to check Sdk.SuppressInstrumentation on start calls and then add the activity id to a list of ignored activities and then remove from that list when stop is called. That way you can ensure that the suppression check only happens on start and properly suppresses only the matching stop events.

Reply:
@reyang thoughts?

Reply:
Sdk.SuppressInstrumentation is a context-aware variable instead of a global variable. I don't think any global flag would work given the nature of the free threading model we have to deal with in the instrumentation library.

Reply:
@reyang do you not think this is an issue overall with the current approach of suppression with incrementing and decrementing a counter? If this is all multi-threaded then you could easily get out of order start and stop calls and end up suppressing the wrong stop call.

Reply:
I don't think the current approach has such issue.

The current approach is a context aware variable https://github.com/open-telemetry/opentelemetry-dotnet/blob/29995ec245748e70293afeeefb41ec38e898c8e4/src/OpenTelemetry/SuppressInstrumentationScope.cs#L29.

It is by default using AsyncLocal, or Remoting CallContext (on legacy runtimes that don't support AsyncLocal).

Reply:
Ok, yeah, that makes sense. So your counter is specific to the async call stack and therefore wouldn't have multi-threading issues with starts and stops.

Well, what is the plan here? Are we going to make those APIs public or are we coming up with something different? I'd really like to get this PR finished up.

Reply:
Personally I think it makes sense to have these functionality moved to the API package (and find better names).
This is a good topic for the SIG meeting.

Reply:
If we promote the Increment/Decrement to public (as done in this PR).
It will force any instrumentations which need to offer "SuppressDownsteam" functionality to take SDK dependency as opposed to API dependency. This means, these instrumentations will only work for *this* particular SDK implementation, and nothing else. If the Supress instrumentation is moved to API, that also doesn't guarantee an alternate SDK implementation will support it, as there is no spec requirement for it, leaving the instrumentations to not work in alternate SDK implementation.

This issue is likely to get more common (as more and more instrumentations gets written), and is a good candidate to pursue a spec for. Once this is spec-ed out, we can expose this in the API (even pursue a change in .NET runtime to support). Until then, leaving this in SDK seems like a good alternate.

@reyang @CodeBlanch @alanwest Please comment if you have objections/alternate proposal. Will check this in next SIG meeting.

Reply:
I guess an alternative is that the the `DiagnosticSourceListener` can have it's own `RuntimeContextSlot` that is incremented when an activity start is skipped and decremented when the corresponding stop is skipped. That would remove the requirement of having these API's (`IncrementIfTriggered` and `DecrementIfTriggered`) made public.

Reply:
@cijothomas I've updated the DiagnosticSourceListener in the contrib project to use this method as an experiment to see if it worked. What do you think about this? https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/a940525eae6c320913d3433e4bc3558c4a01f67f/src/OpenTelemetry.Contrib.Shared/DiagnosticSourceInstrumentation/DiagnosticSourceListener.cs

Reply:
This is also technically same as taking a dependency on *this* SDK, as you are counting on the context entry name. Another alternate SDK implementation may not use the same name, or may not even support this feature.

Reply:
It's technically the same, but it's not requiring you to make additional methods public that you don't seem to want to make public.

Reply:
@cijothomas also, think you missed it, but it's using a different named slot. Not the one that does the internal suppression.

Reply:
I think that approach is reasonable @ejsmith.

Reply:
If we like that approach, then I will update this PR and remove the new public APIs from it. @cijothomas thoughts?

Reply:
thanks @ejsmith for clarifying. I'll come back to this by end of today.
With the recent change to tracerprovider to natively process legacy activities, we may not need the DiagnosticSourceListener to do the Increment/Decrement. Checking that..

Reply:
Pretty sure it's still needed unless something has changed recently. The ES client in the contrib project is already making use of `AddLegacyActivity`.

Reply:
@ejsmith I have checked the most recent code. We need the DiagnosticListener to do the inc/dec counter to achieve this. `TracerProviderSdk` alone won't cover the scenario in the instrumentations. So this(https://github.com/open-telemetry/opentelemetry-dotnet/pull/1856#discussion_r589589947) approach is the best way forward now.

Reply:
Sorry, I correct my comment above. 
As TracerProviderSdk now entirely owns Start/Stop of activity and feeding to processor pipeline, we were able to avoid this check from DSListener as done in this PR: https://github.com/open-telemetry/opentelemetry-dotnet/pull/1893

This should unblock this PR.